graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/ArrayPrototypeBuiltins.java

```diff
@@ -3456,4 +3476,444 @@ protected Object createGroupResult(JSMapObject map, Map<Object, List<Object>> gr
+                        @Cached("createBinaryProfile()") ConditionProfile isJSObject) {
+            checkCompareFunction(comparefn);
+            Object thisJSObj = toObjectOrValidateTypedArray(thisObj);
+            if (isJSObject.profile(JSDynamicObject.isJSDynamicObject(thisJSObj))) {
+                return toSortedJSObject(comparefn, (JSDynamicObject) thisJSObj);
+            } else {
+                return toSortedForeignObject(comparefn, thisJSObj);
+            }
+        }
+
+        private JSDynamicObject toSortedJSObject(final Object comparefn, JSDynamicObject thisJSObj) {
+            long len = getLength(thisJSObj);
+            Object[] array = jsobjectToArray(thisJSObj, len);
+
+            Comparator<Object> comparator = getComparator(thisJSObj, comparefn);
+            if (isTypedArrayImplementation && comparefn == Undefined.instance) {
+                assert comparator == null;
+                prepareForDefaultComparator(array);
+            }
+            sortIntl(comparator, array);
+            reportLoopCount(len);
+
+            JSDynamicObject resultArray = (JSDynamicObject) getArraySpeciesConstructorNode().arrayCreate(len);
+            for (int i = 0; i < array.length; i++) {
+                JSRuntime.createDataPropertyOrThrow(resultArray, Strings.fromLong(i), array[i]);
+            }
+
+            if (isSparse.profile(array.length < len)) {
+                deleteGenericElements(resultArray, array.length, len);
+            }
+            return resultArray;
+        }
+
+        public Object toSortedForeignObject(Object comparefn, Object thisObj) {
+            assert JSGuards.isForeignObject(thisObj);
+            long len = getLength(thisObj);
+
+            if (len >= Integer.MAX_VALUE) {
+                errorBranch.enter();
+                throw Errors.createRangeErrorInvalidArrayLength();
+            }
+
+            Object[] array = foreignArrayToObjectArray(thisObj, (int) len);
+
+            Comparator<Object> comparator = getComparator(thisObj, comparefn);
+            sortIntl(comparator, array);
+            reportLoopCount(len);
+
+            JSDynamicObject resultArray = (JSDynamicObject) getArraySpeciesConstructorNode().arrayCreate(len);
+            for (int i = 0; i < array.length; i++) {
+                JSRuntime.createDataPropertyOrThrow(resultArray, Strings.fromLong(i), array[i]);
+            }
+            return resultArray;
+        }
+
+        private void checkCompareFunction(Object compare) {
+            if (!(compare == Undefined.instance || isCallable(compare))) {
+                errorBranch.enter();
+                throw Errors.createTypeError("The comparison function must be either a function or undefined");
+            }
+        }
+
+        private Comparator<Object> getComparator(Object thisObj, Object compare) {
+            if (compare == Undefined.instance) {
+                noCompareFnBranch.enter();
```

Comment:
(minor) I find it strange you have a branch profile in each branch; we typically have a `ConditionProfile` on the `if` in such cases. But you probably copied that from the original source; interesting we have that pattern there?